### ISSUE TITLE
Extract CSS

### ISSUE THREAD

**Issue by Rich-Harris** [sveltejs/sapper#388]
Currently, Sapper doesn't do anything special with CSS — it expects your bundler to handle it. Which *sort of* works, in that if you don't do anything the CSS will be included as a string inside your JS, and if you configure webpack just right you can extract the CSS by emitting a separate .css asset and importing it from the JS, but it's still a little clunky.

It feels like it ought to be possible to extract the CSS for any components in the app, and make the client-side runtime aware of them. So that when it loads the JS for (say) `/about`, it loads `about-xyz123.css` as well, and creates a link tag. We can avoid any duplication (as currently happens between the critical CSS and the initial route load) while also avoiding FOUC. If we were smart we could probably coalesce blobs of CSS into more optimized chunks.

It *would* mean spelunking through the meta information generated by webpack/Rollup, pulling out things that look like components (based on `.html` extension, presumably?) and pre-compiling them (since this would have to happen prior to bundling) to get the CSS. If someone were using that extension for non-Svelte things, or if they were doing something wacky, there might be creative ways to break it. In a framework like this I *guess* it's okay to have a few constraints though, especially ones that most people will never become aware of?

**Comment by Rich-Harris** [comment]
A plan that might just work: we shove a bunch of placeholder strings into the manifest, like so:

```js
// app/manifest/client.js
const index = {
    js: () => import('../../routes/index.html'),
    css: '__REPLACE_ME_IN_A_TRANSFORMCHUNK_HOOK_XYZ123__';
};

const about = {
    js: () => import('../../routes/about.html'),
    css: '__REPLACE_ME_IN_A_TRANSFORMCHUNK_HOOK_XYZ234__';
};

const blog = {
    js: () => import('../../routes/blog.html'),
    css: '__REPLACE_ME_IN_A_TRANSFORMCHUNK_HOOK_XYZ345__';
};

const blog_ = {
    js: () => import('../../routes/blog/[slug].html'),
    css: '__REPLACE_ME_IN_A_TRANSFORMCHUNK_HOOK_XYZ456__';
};

export const manifest = {
    ignore: [/^\/blog.json$/, /^\/blog\/([^\/]+?).json$/],

    pages: [
        {
            // index.html
            pattern: /^\/?$/,
            parts: [
                { component: index }
            ]
        },
        ...
```

When bundling, we extract information about which components are contained in which chunk, and emit CSS files accordingly. We also note which routes depend on which chunks. Then we simply replace the strings in a `transformChunk` hook, so that it becomes something like this:

```js
// app/manifest/client.js
const index = {
    js: () => import('../../routes/index.html'),
    css: ['xyz123', 'xyz234'];
};
```

Then the runtime can load `/client/css/xyz123.css` and `/client/css/xyz234.css` as necessary. We're already injecting a Rollup plugin to extract information, so that part is straightforward (though it might be necessary to `unshift` rather than `push`, so there's no sourcemap weirdness if the transformation happens after minification). I'm sure the same thing is possible in webpack *somehow*

**Comment by Rich-Harris** [comment]
Argh. I got halfway through implementing this before I realised the fatal flaw: it's not enough to know which components and chunks each route depends on, because we can't actually generate the CSS for each chunk. We *almost* can — we know which components belong to each chunk, and could compile them easily enough — but Sapper can't account for any `preprocess` options that are used in `rollup/client.config.js`, or any subsequent transformations that might exist.

Not sure what the right solution here is. If the Rollup plugin is invoked with `emitCss: true`, Sapper can intercept those imports, store the CSS, and create CSS bundles matching each route later. But it doesn't have access to sourcemaps, because Rollup only passes code to `transform` hooks (because the expectation is that you'll return `{code, map}`, and *it* will take care of composing sourcemaps). So our generated CSS chunks also can't have sourcemaps, which is a dealbreaker for me.

Back to the drawing board.
